### DISCUSSION TITLE
How can I pin plugins from my user configuration?

### DISCUSSION THREAD

**Discussion by danielo515** [LunarVim/LunarVim#2914] [Q&A] [answered]
Hello.
I was wondering if it is possible to set my own pinned configuration for the plugins I use? I am usually happy as they are when I install them, and generally when they introduce something new is to break my configuration.
I would love to have a packer command to run to create a snapshot of the currently installed plugins (excluding lvim core ones) and don't let packer install anything newer than that.I had the impression that this is somehow supported, but I can't figure out how

**Comment by rebuilt** [comment, marked as answer]
In your plugin configuration set the `commit` option to the commit SHA you'd like to use.  You can get the SHA by running `git log` in the git folder of the plugin you're trying to pin.   The SHA will look like like this: ` 5ccdf6a88d5790382ad3da6bb908c606765754e0`

Here's an example of what your packer config should end up looking like:   https://github.com/LunarVim/nvim-basic-ide/blob/master/lua/user/plugins.lua

**Reply by latipun7** [reply]
I'm guessing he (and me) want to get the auto something, the snapshot file just for the user defined plugins, just like the lvim's snapshot for core plugins.

**Reply by rebuilt** [reply]
Ah, that's cool.  I didn't know about this Packer feature.  Here is the commit that added that functionality.  https://github.com/wbthomason/packer.nvim/pull/370

The packer commands `:PackerSnapshot <snapshot_name>` and `:PackerSnapshotRollback <snapshot_name>` should operate on all plugins.   

I don't know of a way to exclude the LunarVim builtin plugins from the snapshot.   

It looks like a `snapshot` is not the same thing as pinning to a commit.   Running a `:PackerSync` will still update plugin versions.  But if anything breaks you can `:PackerSnapshotRollback` to an earlier snapshot.

**Reply by latipun7** [reply]
Yeah, if pinning it to commit, I'm doing it to pin to the latest tag of some custom plugin I add (`tag = "*"`)

**Reply by danielo515** [reply]
Ok, that is new information I didn't knew. I was expecting that, having an existing snapshot packer will never update the plugins that are pinned within the snapshot. Seems that my best option is to create a snapshot and then cherry pick the commits I want to keep and put them in my plugin configuration.

**Reply by latipun7** [reply]
That's what snapshot in packer doing, I guess. So you could rollback to latest snapshot if something broke.
But in lvim plugin configs, lvim's snapshot is something like pinning the commit since the code looks like this (just my guess):
https://github.com/LunarVim/LunarVim/blob/45f9825d1e666890ed37baf15a14707ae40e5cff/lua/lvim/plugins.lua#L267
basically, it is doing packer config with `commit = commitsha` option on all plugins, but they do it "kind of" automatically.

**Reply by danielo515** [reply]
Oh, that is quite smart. I'll give it a go, because it is better than manual cherry pick 😄

**Reply by danielo515** [reply]
I finished a custom solution for my use-case.
Will a PR to add this functionality to lunarvim be appreciated ? Something that finds a packer-snapshot.json file in the runtime path and then pins all the user plugins to the commit specified there if there is any?
@kylo252 what do you (and any other) think about it?
